taichi/program/async_engine.cpp

```diff
@@ -130,25 +131,35 @@ void AsyncEngine::launch(Kernel *kernel) {
     kernel->lower(/*to_executable=*/false);
   auto block = dynamic_cast<Block *>(kernel->ir.get());
   TI_ASSERT(block);
+
   auto &offloads = block->statements;
-  auto &dummy_root = kernel_to_dummy_roots_[kernel];
-  if (dummy_root == nullptr) {
-    dummy_root = std::make_unique<Block>();
-    dummy_root->kernel = kernel;
+  auto &kmeta = kernel_metas_[kernel];
+  const bool kmeta_inited = kmeta.initialized();
+  if (!kmeta_inited) {
+    kmeta.dummy_root = std::make_unique<Block>();
+    kmeta.dummy_root->kernel = kernel;
   }
   for (std::size_t i = 0; i < offloads.size(); i++) {
     auto offload = offloads[i]->as<OffloadedStmt>();
-    KernelLaunchRecord rec(
-        kernel->program.get_context(), kernel,
-        clone_offloaded_task(offload, kernel, dummy_root.get()));
+    auto cloned = clone_offloaded_task(offload, kernel, kmeta.dummy_root.get());
+    uint64 h;
+    if (kmeta_inited) {
+      h = kmeta.offloaded_hashes[i];
+    } else {
+      h = hash(cloned.get());
+      TI_ASSERT(kmeta.offloaded_hashes.size() == i);
```

Comment:
What if we have two same offloads in a kernel? And does it prevent two different offloads in two different kernels having the same hash?

Reply:
BTW if the hash really crashes, I think it should be an error rather than an assertion.

Reply:
> BTW if the hash really crashes, I think it should be an error rather than an assertion.

Note that `offloaded_hashes` is a `std::vector`. It maps to an offloaded task by its *position* in that kernel... The assertion is checking for position (i.e. the hash of this offloaded task isn't computed yet), not hash collision.

> What if we have two same offloads in a kernel? And does it prevent two different offloads in two different kernels having the same hash?

I actually think it's fine for the same offloads (regardless of whether they are in the same kernel or not) to have the same hashes. If two offloads are indeed the same, then they should have the same `TaskMeta`s (i.e. input/output snodes, activation snodes)?

Reply:
Oh I see. I thought `offloaded_hashes` was an `std::unordered_map`, which would cause an assertion failure when there are two same offloads...

> I actually think it's fine for the same offloads to have the same hashes.

I agree.